VerilogForVerificationStyle.md

```diff
@@ -0,0 +1,826 @@
+| Scoreboard                        | `<dut>_scoreboard scoreboard;`               |
+| Virtual sequencer                 | `<dut>_virtual_sequencer virtual_sequencer;` |
+| Sequencer                         | `<dut>_sequencer <dut>_sequencer;`           |
+
+The UVM testbench generation script will automatically create skeletons of these
+components with this naming scheme, so almost all of this naming is automated
+out.
+Additional naming guidelines:
+
+1. When instantiating any objects, the `name` argument to `type_id::create()`
+must match the name of the variable being assigned to.
+This ensures that any reporting messages refer to the right objects.
+2. To avoid name collisions, it is recommended that global types and types defined
+within a package have unique prefixes. This includes typedefs, enumerated
+types/values, package parameters, package localparams, any functions/tasks, and
+macros. Note that all package variables (parameters, etc.) must have a specified type.
+3. When instantiating a user-defined class member variable that is a class object:
+    *   Prefix the instance name of the class object with `m_`, as such:
+
+```systemverilog
+my_animal m_my_animal;
+```
+
+This scheme does not apply to any class member variables that are
+not themselves class objects.
+
+* If multiple instantiations of the class are desired without an array
+to store them, use `<class_name> m_<instance_name> to declare a variable,
+and use an informative instance name.
+
+```systemverilog
+my_animal m_my_cat;
+my_animal m_my_dog;
+my_animal m_my_fish;
+```
+
+* If an array of class handles is desired, choose a name such that it follows
+one of the following schemes:
+
+```systemverilog
+int NUM_ANIMALS = 10;
+// This plural convention is recommended, but both are equally acceptable
+my_animal my_animals[NUM_ANIMALS];
+my_animal my_animal_array[NUM_ANIMALS];
+```
+
+
+To avoid name collisions, it is recommended that global types and types defined
+within a package have unique prefixes. This include typedefs, enumerated
+types/values, package parameters, package localparams, any functions/tasks, and
+macros.
+This prefixing is not necessary for any types defined within a class, as they
+will be scoped by the class.
+
+## Directory Structure and Packages
+
+*   Define one class per file, and the filename should match the name of the
+    class. The class files must be `` `include``-ed into the package file.
+*   Directory structure of all relevant testbench files should follow the
+example below:
+
+```systemverilog
+<block>_agent/
+  <block>_if.sv
+  <block>_item.sv
+  <block>_agent.sv
+  <block>_agent_pkg.sv
+  <block>_agent_cfg.sv
+  <block>_agent_cov.sv
+  <block>_driver.sv
+  <block>_host_driver.sv (Only if block requires a separate driver for host mode)
+  <block>_device_driver.sv (Only if block requires a separate driver for device mode)
+  <block>_monitor.sv
+  <block>_sequencer.sv
+  seq_lib/
+    <block>_base_seq.sv
+.hw/ip/<block>/dv/
```

Comment:
Can `<block>_agent` and `hw/ip/<block>/dv` have different names. One IP (dut) may require multiple agents

Reply:
Block agent and block env should be two separate topics / headings. They are not necessarily tied to each other (as Tung pointed out; plus, some blocks dont even have an agent). 

Also better to not make assumption on block env's directory path. It should be enough to indicate that it should reside alongside the DUT RTL code.

Reply:
Maybe not call it 'block' agent - it can be any agent not necessarily tied to a particular block. Just call it 'interface' agent?

Reply:
That is a good point, agree that the agent directory should be more tied to the interface that it is built around than the block in question, I'll update the naming to `<interface>_agent/` and split off the agent and env into two separate sections.

Reply:
Agree, I think keeping the <block>_agent/ scheme and changing to <dut>_env would be the best option.